### ISSUE TITLE
Update Wiki with merge strategies

### ISSUE THREAD

**Issue by RussKie** [gitextensions/gitextensions#4219]
Wiki needs an update to provide a clear expectations around hygiene and handling of commits.

From @RussKie https://github.com/gitextensions/gitextensions/pull/4210#issuecomment-350538677

> I am of an opinion that a PR should consist of a single commit which contains a single self-contained atomic change. That can be easily reasoned about or that can be reverted, if necessary.
On occasions a PR may contain additional non-functional commits which are necessary or complimentary to the change. E.g.
    _commit1: update packages (non functional)
    commit2: make the real change (functional)
    commit3: code maintenance/refactor (non functional)_
>
> As a reviewer or a fellow developer I don't really care that it took another developer 57 commits to deliver his/her change, that it fixed typos or responded to reviews 29 times, reworked the implementation 22 times and fixed unit tests 6 times. All I care the change is done and delivered.
Personally I find it easier to review a single commit too then hunt for a change across multiple commits.

From @gerhardol https://github.com/gitextensions/gitextensions/pull/4210#issuecomment-350586959

> I would like to see the position on merging from master vs rebase (with force push) both when there are conflicting changes in master, just related or just to minimize the diff to master (to not have the maintainer doing that diff).
My opinion: Force push should never be a part of a workflow, it is evil to change the history.
>
> Similarly, I would like to see the maintainer instructions on merging vs squashing vs rebasing the PRs. (The answer may well be "it is up to the maintainer").
>
> Another item is how to handle "enablers" and dependent PRs. I for instance handled #4134 as a separate PR where the real change were in #4157. This was done to simplify reviews. But #4134 in itself just added dead code. I tried to explain that, but there are over 30 PRs related to #4031.
I have tried to make separate PRs but the lead time does get very long with a chain of maybe 5 chained PRs and I could maybe have split more . (I rebased maybe 30 times to prepare for #4157).
>
> For reviewing: I would like the maintainers to encourage "voting" on issues and PRs: Reporting that a change seem reasonable should give better confidence and trying it out increases more. Users should try to increase their "standings" with testing to get their "pet" changes reviewed and maybe fixed.

@all: Opinions? Thoughts? Feedback?

**Comment by jbialobr** [comment]
> I am of an opinion that a PR should consist of a single commit which contains a single self-contained atomic change. That can be easily reasoned about or that can be reverted, if necessary.

I don't like it. It is very unlikely that any non-trivial change can be done as a single **atomic** commit. 
See https://github.com/gitextensions/gitextensions/pull/3835/commits/58997e7cdd5df9f998f2426e8b92ade8b593aed6 for example. I would not call it an atomic commit. The commit message is 6 screens long!
It is important to have atomic changes to be able to reason about the causes and needs that change was introduced for. And it is needed after some time has passed and you find a bug or need to change sth.
Then instead of knowing the real reason you will only get a message like: "This change implements feature  X". 
As a reviewer I review changes that the merge commit introduces. The intermediate commits may help to know the context of the changes and I think that the more the merrier.

**Comment by vbjay** [comment]
I agree. The history is important. Knowing how they got there is helpful.

**Comment by gpongelli** [comment]
I think that dependent PR can be a managing hell and git/github are not so smart in this way because merge commit on master don’t track this dependency (how do you identify commit to be revented in the case?).

About squash, isn’t possible to set “default auto-squash” when merging PR? This reduces developer effort that has to squash manually (and you’ve asked me to squash many time during that GPG changes).
Gitlab has this option, so maybe also github has.

Other point is rebase, the maintainer should ask a PR rebase only after all the review are ended, due to the history rewriting. This because github marks review as “old” if it appears on a commit that no more exists (due to the rebase).

My PR can be an example on how things could be done, in that commit there is everything, from code to test to resources, so reverting it you remove everything without the needs to search other changes in different PR.
Obviously a PR ca accept PR like we did, so the feature or the change can be developed from zero to 100%, scoped in that PR.

Ps
My PR about showing GPG info is too long because it was developed in three different way before arriving at the final one. Meanwhile many rebase due to new effort on master branch and squash due many commit I did. Discussion was around 200 comment, and I don’t know how many commit I did ;)

My 2 cents.

**Comment by gerhardol** [comment]
Of course, before submitting PRs, branches should be rebased and squashed so they are easy to review.
I prefer to merge rather than squash to preserve the commit information in the case the PRs was accepted with few changes.

The question is still for noisy PRs and where master is changed: What should the requester and the maintainer do?

**Comment by jbialobr** [comment]
@sharwell @drewnoakes 
Would appreciate your opinion based on your experience.

**Comment by drewnoakes** [comment]
I had an interesting discussion about this last night actually, though in the more general context of what it takes to maintain quality of product and an inclusive community in an open source setting. Good timing.

That more general context is interesting too, and a direction there informs the decision here.

But, on topic. There are many things to trade off with respect to squashing PR commits, and identifying priorities between then is probably a good place to start. Here are the main ones I can think of, in no particular order:

1. Friction for contributors
2. Quality of resulting code
3. Ability to identify and fix bugs that slip through review
4. Chance that a PR goes stale due to conflicts
5. Ability to extract information from the commit history

In my opinion, asking contributors to squash their PRs:

1. adds more friction for contributors
2. doesn't improve the quality of the end result, and may actually decrease it, if conflict resolution introduces bugs
3. reduces ability to detect historical bugs because bisecting is likely to hit larger chunks
4. increases the time before a PR is merged, depending upon how responsive/patient a contributor is
5. depends upon the nature of the squashing

---

On point 5, if a PR has commits resembling:

- add feature
- more stuff
- fix bug
- change thing
- undo
- fix tests
- add docs
- fix build

Then to me this is a signal that this contributor may be new to PRs and would benefit from some support from more experienced maintainers. This particular sequence could very likely be squashed into one commit, and it's probably that the contributor doesn't know how to do that. Direction to documentation explaining what a nice commit history might look like, and how to rebase interactively, or how to reset and stage in patches (using GE of course), and why that's a really cool part of git... that's a likely a positive experience with a positive outcome for everyone.

---

PRs that add a feature often resemble:

- Expose Foo.Bar to support thingies
- Add ThingyForm
- Add buttons to show ThingyForm 

Here the commits tell you something about what the feature requires, what it's comprised of and how it integrates. Reviewing this would likely be easy, and leaving the commits in the history, along with a merge commit that points to the GitHub PR for further discussion, allows access to the full context and thinking around the change.

---

PRs that refactor are a bit trickier. I've been doing a lot of those lately. As they involve nearly as many deletions as insertions, they're hard to review unless each commit is focussed in purpose, obvious from the commit message, and small in scope. Squashing these results in something that's very hard to come back to later.

---

@jbialobr pointed out 58997e7 as an example of a commit with a very long commit message. Looking at the diff though, it's mostly additions. While I would probably merge that as is, you could argue that those sub-commits were the author's inner working and not relevant to code that existed before the PR was merged. So maybe the number of additions/deletions is something to factor in to a heuristic.

I agree with @sharwell's exceptions to when editing PR history makes sense. His guidelines around asking the for contributor's permission to edit their work is respectful and inclusive as well.

---

I worked on the NetMQ project, which is part of ZeroMQ. They have a a very interesting [formal process](https://rfc.zeromq.org/spec:22/C4/) for contributions and project management. I was shocked to see that it actually worked -- the project lead merged almost every PR without much friction at all. Other members of the team would submit follow up PRs for issues that were uncovered. The philosophy there was to include all activity, then lock things down for a period of testing and refinement prior to release. I'm not saying that approach is suited everywhere, but it totally surprised me that it worked at all, let alone well, and I have since adjusted my fear of accepting PRs that aren't perfect, for some definition of perfect.

---

Having a set of standards documented so that contributors know what's expected of them is a great idea. I read through `CONTRIBUTING.md` and came away with a general sense, but there's much more that could be captured there. Without such documentation the review process can appear arbitrary which would likely discourage contributors from becoming maximally involved. I'm not saying it is arbitrary, but it could seem that way. My early contributions received a lot of style-oriented feedback which was a little frustrating as there was no guideline documented, and the existing code wasn't representative of what was expected. Introducing StyleCop.Analyzers has improved that situation by having some of the project's policies executing on the developer's computer, saving reviewers time and reducing back and forth during review. There's more we can do, both with additional documentation and Roslyn analysers.

---

Anyway, that's what comes to mind. Sorry it's so long. And I hope as a newcomer around here I haven't spoken out of turn. I think this is a good community and the fact that these discussions are happening this way are a healthy sign.

**Comment by gerhardol** [comment]
@drewnoakes 
Nice summary

I believe the order of importance to keep a project like GE successful is:
 1. Keep the architects/primary maintainers happy (that is RussKie, jbialobr).
 2. Keep contributors productive.
 3. Satisfy users needs. Last priority, but if 1 and 2 works well, this follows.

After a review I prefer to not see a squash or rebase as it slows down reviews - you do not know what has been changed. Merges are fine.
There are exceptions, With many commits and merges, it will be hard to review the complete PR, then it may be OK to ask for a squash.

> On point 5, if a PR has commits resembling:
> Then to me this is a signal that this contributor may be new to PRs

Quite often, PRs look like the example after reviews. Squashing (interactively) is very relevant in these situations.

**Comment by jbialobr** [comment]
>I believe the order of importance to keep a project like GE successful is:

Perhaps this is a subject for another interesting discussion, but for sure I would exclude myself from the list of architects/primary maintainers. Never wanted be a one - just happened that there was no one around for some time, I am just happy with contributing here and there, learning sth new. I am happy that GE drew new members to the team.

**Comment by drewnoakes** [comment]
> I believe the order of importance to keep a project like GE successful is: Keep the architects/primary maintainers happy

You're right. I wasn't considering the mental state of the project team. Obviously there's no project without them, so they have to be happy and operating well.

I noticed another minor thing about squashing today. PR #4734 was squashed, which contained three commits:

* Replace DeconstructionExtensions Tuple methods with System.ValueTuple
* Replace usages of Tuple with ValueTuple
* Remove redundant using directives

The last commit was only tangentially related. Removing `Tuple` meant that a bunch of `using System;` directives became redundant. I tidied the whole solution automatically, which touched some other files.

The end result is that the current commit message for several files (`EnvironmentConfiguration.cs`, `GitModule.cs`, `AssemblyInfo.cs`, `Program.cs`, `FormBisect.cs` and so on...) is currently "Replace Tuple with ValueTuple". These files likely never used `Tuple<...>`. Instead the more apt "Remove redundant using directives" comment would have applied if the merge hadn't been squashed.

Similarly, the commit history for those files now includes a message about tuples that doesn't match their diff.

I expect folks to say the answer is to split such changes out into separate PRs. I would probably not have created a PR just to tidy up using directives. There may be examples of this that are more fuzzy.

**Comment by jbialobr** [comment]
>I expect folks to say the answer is to split such changes out into separate PRs.

I believe there is no a single answer for that. Depending on the circumstances it may be: move to a new PR, squash, amend to other commit, do not squash (leave as is). I think it is safe to follow the @sharwell's policy:
>In the event a PR needs to be rewritten, I always ask for permission from the author before doing so, and explain why it was necessary for that pull request.

We should let contributors to say their opinion, perhaps there is a reason that would cause the maintainer to change his mind.

**Comment by RussKie** [comment]
An interesting discussion with some good points. 

Where I work we follow SR principles - that is a commit to contain a self-contained atomic change. This requirement typically makes commits smaller, easier to review and reason about. We have a lot of engineers (onshore and offshore) with different levels of skills and experience, and following SR principles creates a more uniform environment.
We do consider multi-commit PRs, however functional and non-functional changes are to be separated.

Of course not everything can be done with small PRs - like #4601 or #3693  or #3835. However in these large PRs there are commits which arguably contain iterations of internal workings and can be be squashed (e.g. PR#1 - add some method, PR#4 - changed the method, PR#8 - renamed the method, PR#13 - removed method etc). 
Also some of PRs contains lots of commits addressing PR comments - e.g. #3835 contained many dozens of commits. The net value of those to me personally as a contributor is minimal - I am primarily concerned with the net result, especially if the diff is easy to follow.

I certainly appreciate the value of multi-commit PRs - they can make reviewers life significantly easier, especially if diff is destructive, or there are series of refactors.
However I find it much harder to review PRs with dozens of commits, especially if they span a period of time....

**Comment by jbialobr** [comment]
From @drewnoakes https://github.com/gitextensions/gitextensions/issues/4754#issuecomment-377880994

>I'm glad the PR wasn't completely squashed as bisecting will narrow down the problem to a smaller changeset.
